cyral/resource_cyral_integration_idp_saml_draft.go

```diff
@@ -0,0 +1,281 @@
+		},
+		NewResourceData: func() ResourceData { return &CreateGenericSAMLDraftRequest{} },
+		NewResponseData: func() ResponseData { return &GenericSAMLDraftResponse{} },
+	}
+}
+
+func ReadGenericSAMLDraftConfig() ResourceOperationConfig {
+	return ResourceOperationConfig{
+		Name:       "GenericSAMLDraftResourceRead",
+		HttpMethod: http.MethodGet,
+		CreateURL: func(d *schema.ResourceData, c *client.Client) string {
+			return fmt.Sprintf("https://%s/v1/integrations/generic-saml/drafts/%s", c.ControlPlane, d.Id())
+		},
+		HandleRequestError: func(d *schema.ResourceData, c *client.Client, err error) error {
+			httpError, ok := err.(*client.HttpError)
+			if !ok || httpError.StatusCode != http.StatusNotFound {
+				return err
+			}
+			log.Printf("[DEBUG] SAML draft not found. Checking if completed draft exists.")
+
+			query := urlQuery(map[string]string{
+				"includeCompletedDrafts": "true",
+				"displayName":            d.Get("display_name").(string),
+				"idpType":                d.Get("idp_type").(string),
+			})
+			url := fmt.Sprintf("https://%s/v1/integrations/generic-saml/drafts%s",
+				c.ControlPlane, query)
+			body, err := c.DoRequest(url, http.MethodGet, nil)
+			if err != nil {
+				return fmt.Errorf("unable to read completed drafts: %w", err)
+			}
+
+			resp := ListGenericSAMLDraftsResponse{}
+			if err := json.Unmarshal(body, &resp); err != nil {
+				return fmt.Errorf("when reading completed drafts: "+
+					"unable to unmarshall JSON: %w", err)
+			}
+
+			myID := d.Id()
+			found := false
+			for _, draft := range resp.Drafts {
+				if draft.ID == myID {
+					found = true
+					break
+				}
+			}
+			if !found {
+				log.Printf("[DEBUG] Completed draft with ID %q "+
+					"not found. Triggering recreation.", myID)
+				currentToggle := d.Get("toggle_recreation").(bool)
+				newToggle := !currentToggle
+				if err := d.Set("toggle_recreation", newToggle); err != nil {
+					return fmt.Errorf("error forcing recreation of SAML draft: %w", err)
+				}
+			} else {
+				log.Printf("[DEBUG] Found completed draft with ID "+
+					"%q.", myID)
+			}
+			return nil
+		},
+		NewResponseData: func() ResponseData { return &GenericSAMLDraftResponse{} },
+	}
+}
+
+func DeleteGenericSAMLDraftConfig() ResourceOperationConfig {
+	return ResourceOperationConfig{
+		Name:       "GenericSAMLDraftResourceDelete",
+		HttpMethod: http.MethodDelete,
+		CreateURL: func(d *schema.ResourceData, c *client.Client) string {
+			return fmt.Sprintf("https://%s/v1/integrations/generic-saml/drafts/%s", c.ControlPlane, d.Id())
+		},
+		HandleRequestError: func(d *schema.ResourceData, c *client.Client, err error) error {
+			httpError, ok := err.(*client.HttpError)
+			if !ok || httpError.StatusCode != http.StatusNotFound {
+				return err
+			}
+			log.Printf("[DEBUG] SAML draft not found. Skipping deletion.")
+			// If the HTTP status is NotFound, it means we don't need to
+			// worry about deleting anything.
+			return nil
```

Comment:
I see. The problem I see though is that, for example in our case, we need to check if the draft still exists in the `completed` state. If we trigger a recreation every time `404` is returned, we would effectively trigger recreation every time `terraform apply` is run if there is a `cyral_integration_idp_saml` resource linked to the draft. Both the draft and the integration itself would have to be recreated. Having a fully custom function to handle the error makes more sense to me. We could construct models of common error-handling function patterns we see in the future, to avoid code duplication.